new Node.js demo for batch geocoding a .csv

mpayson opened the pull request:
Hey team, first take on a batch geocode sample for Node, would love feedback on:

* Is best batch geocode sample a Node script?
* General logic/structure/style/formatting (my first PR!)
* ~Excel thinks the output `.csv` is corrupt, not sure why or how to fix?~ might be local problem

@phpmaps @jgravois

jgravois commented:
very, _very_ cool @mpayson!

i've got other fires blazing now, but i'll find time to review and help you get this landed soon.

jgravois commented on demos/batch-geocoder/README.md:
```diff
@@ -0,0 +1,7 @@
+# Running this demo
+
+1. `$ git clone https://github.com/mpayson/arcgis-rest-js.git`
```
all the other demos assume the repo has already been cloned. 

nothing wrong with you deviating, but you should point folks at Esri/arcgis-rest-js.

jgravois commented on demos/batch-geocoder/README.md:
```diff
@@ -0,0 +1,7 @@
+
+1. `$ git clone https://github.com/mpayson/arcgis-rest-js.git`
+2. `$ cd arcgis-rest-js/demos/batch-geocoder/`
+3. `$ npm install`
```
since this is a monorepo, we shouldn't direct them to `npm install` inside individual packages.

instead they should run `npm run bootstrap` in the root of the repo to appropriately symlink the needed dependencies.

jgravois commented on demos/batch-geocoder/batch-geocode.js:
```diff
@@ -0,0 +1,109 @@
+      resultMap = mapResults(res);
+      output = data.map((row, i) => {
+        id = row[config.fieldmap.OBJECTID] || i;
+        return {...row, ...resultMap[id]};
```
since you're using spread operators and other fancy new JS stuff, you need to make a point in the doc of explaining that Node.js 8.x or higher is required.

mpayson replied:
:white_check_mark: added dependency to [readme](https://github.com/Esri/arcgis-rest-js/pull/105/files#diff-6aad8c67c6e25840155d92da311f8f16R3).

I think spread etc only [requires >6.x](http://node.green/)? Can also update to support older versions?

jgravois commented on demos/batch-geocoder/config-template.js:
```diff
@@ -0,0 +1,6 @@
+  un: "<USERNAME>",
+  pw: "<PASSWORD>",
+  csv: "<PATH/TO/CSV>",
+  fieldmap: "<CSV TO REQUEST FIELD MAP OR ADDRESS FIELD>"
```
i'd recommend:

1. encasing `un` and your other properties as strings
2. just hardcoding the .csv path to use your example file.
3. either doing the same with the fieldmap or at least including it in copy/pasteable form in a code comment in the same file.

```json
{ 
  "fieldmap": {
    "postal": "ZIPCODE"
  }
}
```
in general, the fieldmap was the biggest point of confusion when i ran your app. it'd make this sample substantially more self explanatory if you talked through what it is and what it is doing.

jgravois commented on demos/batch-geocoder/package.json:
```diff
@@ -0,0 +1,36 @@
+{
+  "name": "batch-geocoder",
+  "version": "0.0.1",
```
use the same version number as everything else in the repo. `1.0.2`.

jgravois commented on demos/batch-geocoder/package.json:
```diff
@@ -0,0 +1,36 @@
+    "arcgis-rest-js"
+  ],
+  "author": "Max Payson (mpayson)",
+  "license": "ISC",
```
"Apache-2.0"

jgravois commented on demos/batch-geocoder/package.json:
```diff
@@ -0,0 +1,36 @@
+    "@esri/arcgis-rest-request": "^1.0.2",
+    "isomorphic-fetch": "^2.2.1",
+    "isomorphic-form-data": "^1.0.0",
+    "papaparse": "^4.3.6"
```
you should include "private": true to ensure this doesn't accidentally get published to npm.

mpayson replied:
✅ [here](https://github.com/Esri/arcgis-rest-js/pull/105/files#diff-326f028d30a88b72b4942c64c32515e3R23)

jgravois commented on demos/batch-geocoder/package.json:
```diff
@@ -0,0 +1,36 @@
+{
+  "name": "batch-geocoder",
+  "version": "0.0.1",
+  "description": "arcgis-res-js batch geocode sample",
```
arcgis-res-js > arcgis-res**t**-js (with a 't')

jgravois commented:
at some point, it'd be worth considering using something like [Sanitize.js](https://github.com/gbirke/Sanitize.js/blob/master/README.md) or even a simple regex to remove <script> tags from within values inside the .csv (in the unlikely event of an [`XSS`](https://www.owasp.org/index.php/Cross-site_Scripting_(XSS))).

jgravois commented:
are you going to have a chance in the next two weeks to take one more pass at this? i'd like to show it off at DevSummit so i can get it landed if push comes to shove.

mpayson commented:
@jgravois , yes, I'll have it done by end of week!

Side note, should we include it as part of [Node CLI](#112 ) example? Then doc in a `readme` for people looking at batch-geocoding/search/export samples?

mpayson commented:
@jgravois 
* [x] now assumed cloned
* [x] added node dependency to [readme](https://github.com/Esri/arcgis-rest-js/pull/105/files#diff-6aad8c67c6e25840155d92da311f8f16R3). I think spread etc only [requires >6.x](http://node.green/)? Can also update to support older versions?
* [x] config updates
  * [x] encased `un` and other properties as strings, and described them in [readme](https://github.com/Esri/arcgis-rest-js/pull/105/files?diff=unified#diff-6aad8c67c6e25840155d92da311f8f16R6)
  * [x] hardcoded CSV path and the `fieldmap`
  * [ ] ❓ added quick `fieldmap` explanation in readme and added options as [comments](https://github.com/Esri/arcgis-rest-js/pull/105/files?diff=unified#diff-f81fb4e155d77d7b7dc98a65b5fd59aaR8) in `config.js`. thoughts? agreed this is confusing.
* [x] `"private": true` [here](https://github.com/Esri/arcgis-rest-js/pull/105/files#diff-326f028d30a88b72b4942c64c32515e3R23)
* [ ] remove scripts tags

jgravois commented:
> should we include it as part of Node CLI example #112?

nah, lets keep them separate. i'd like to see our collection of demos grow to include lots of different browser samples and lots of Node.js samples too.

jgravois commented:
pushed one more commit to rename the directory `batch-geocoder-node` and append the context to a few others.

thx again @mpayson for contributing this sweet demo!
